identity_stardust/src/did/stardust_did.rs

```diff
@@ -0,0 +1,927 @@
+  ///
+  /// # Errors
+  ///
+  /// Returns `Err` if the input is not a syntactically valid [`StardustDID`].
+  pub fn check_validity<D: DID>(did: &D) -> Result<()> {
+    Self::check_method(did).and_then(|_| Self::check_method_id(did))
+  }
+
+  /// Checks if the given `DID` has a valid [`StardustDID`] `method` (i.e. `"stardust"`).
+  ///
+  /// # Errors
+  ///
+  /// Returns `Err` if the input represents another method.
+  // TODO: Change the naming in the docs once we remove the code for the current IOTA method.
+  pub fn check_method<D: DID>(did: &D) -> Result<()> {
+    (did.method() == Self::METHOD)
+      .then_some(())
+      .ok_or(DIDError::InvalidMethodName)
+  }
+
+  /// Checks if the given `DID` has a valid [`StardustDID`] `method_id`.
+  ///
+  /// # Errors
+  ///
+  /// Returns `Err` if the input does not have a [`StardustDID`] compliant method id.
+  // TODO: Is it correct to also validate the network here? The current IOTA DID method does NOT do that.
+  pub fn check_method_id<D: DID>(did: &D) -> Result<()> {
+    let (network, tag) = Self::denormalized_components(did.method_id());
+    NetworkName::validate_network_name(network)
+      .map_err(|_| DIDError::InvalidMethodId)
+      .and_then(|_| Self::check_tag_str(tag))
+  }
+
+  /// Returns a `bool` indicating if the given `DID` is valid according to the
+  /// [`StardustDID`] method specification.
+  ///
+  /// Equivalent to `Self::check_validity(did).is_ok()`.
+  pub fn is_valid(did: &CoreDID) -> bool {
+    Self::check_validity(did).is_ok()
+  }
+
+  /// Returns the Tangle `network` name of the `DID`.
+  pub fn network_str(&self) -> &str {
+    Self::denormalized_components(self.method_id()).0
+  }
+
+  // foo:bar -> (foo,bar)
+  // foo -> (NetworkName::DEFAULT_STR, foo)
+  #[inline(always)]
+  fn denormalized_components(input: &str) -> (&str, &str) {
+    input
+      .find(':')
+      .map(|idx| input.split_at(idx))
+      .map(|(network, tail)| (network, &tail[1..]))
+      .unwrap_or((NetworkName::DEFAULT_STR, input))
+  }
+
+  /// Returns the unique Tangle tag of the `DID`.
+  pub fn tag(&self) -> &str {
+    Self::denormalized_components(self.method_id()).1
+  }
+
+  #[cfg(feature = "alias-id")]
+  /// Returns the [`AliasId`] corresponding to this [`StardustDID`].
+  pub fn alias_id(&self) -> AliasId {
+    AliasId::from_str(self.tag())
+      .unwrap_or_else(|_| panic!("A {} DID should always provide a valid AliasId", StardustDID::METHOD))
+  }
+
+  /// Set the network name of this [`StardustDID`].
+  pub fn set_network_name(&mut self, name: NetworkName) {
+    let new_method_id: String = format!("{}:{}", name, self.tag());
+    // unwrap is fine as we are only replacing the network
+    self.0.set_method_id(new_method_id).unwrap();
+  }
+
+  /// Replace the network name of this [`StardustDID`] leaving all other segments (did, method, tag) intact.  
+  pub fn with_network_name(mut self, name: NetworkName) -> Self {
+    self.set_network_name(name);
+    self
+  }
```

Comment:
`IotaDID` and `StardustDID` are not meant to be mutable in general, and I can't think of a use-case where changing the network of an existing IOTA DID would be necessary.

Leaving `with_network_name` might be fine.

```suggestion
  /// Replace the network name of this [`StardustDID`] leaving all other segments (did, method, tag) intact.  
  pub fn with_network_name(mut self, name: NetworkName) -> Self {
    self.set_network_name(name);
    self
  }
```